# pipeline_annotation
Annotate genomes and format them to gembase format.  

**Input:**
- File with list of genomes to format: 1 genome per line, 2 columns, no header. First column with gembase name of the genome, 2nd column with the current name of the genome (with file extension). The gembase name is: GGSS.mmyy.nnnnn with GGSS the 2 first letters of gender and 2 first letters of species, mmyy the month and year, nnnnn the strain number (with trainling 0s if needed).
- all genome sequences in multifasta (with current name) in a same folder `dbpath`.

**Output:**
- A folder `respath` with subfolders `Genes`, `Replicons`, `Proteins`, `LSTINFO` with all genomes in each.
- The list of genomes updated with 2 new columns: number of contigs and total length of sequence
- The log files .err and .out

## Prepare sequences
Prepares the raw genomic sequence files for the pipeline, with `prepare_sequences.py`:
- rename all contigs of a given multifasta file. Contigs are named with the genome name (= fasta filename) + the contig number + the size of the contig (bp)
- get the total number of contigs in each genome
- get the total size (in nuc) o each genome

Input: 
- database folder, containing all sequences in multifasta, with original names
- a file, "lstinfo-file" with 2 columns: gembase name and original name of each genome

Output: 
- for each genome, a new multi-fasta in the database folder (with contig names changed), called "original_name'-gembase.fna
- a file "lstinfo-file"-complete.lst with 4 columns: gembase name, original name, nb contigs, genome size

## Annotate genomes
used :
prokka 1.11 (with hmmer/3.1b1 aragorn/1.2.36 barrnap/0.4.2 minced/0.1.6 blast+/2.2.28 prodigal/2.60 infernal/1.1 ncbi_toolbox/20151127 signalp/4.0)

- Script `prokka_array.sh`
- Output: In the `Database` folder, for each genome, a subfolder `<genome_original_name>-prokka11Res` with prokka output files (see https://github.com/tseemann/prokka#output-files).

To check that Prokka ran well on all genomes: `check_prokka_run.sh` (called from `prokka_array.sh`). If problems occur, they are added to the error output file, called `err-<LSTINFO_file>-<current_date>.err`.


## Translate outputs to gembase format

This is done with the `generate_gembase.sh` script (called from `prokka_array.sh`):
- Parse .tbl prokka results and transform to .lst, with one line per gene, with the following fields: start, end, strand (C or D), type (CDS, RNA, CRISPR...), gembase_name, gene_name | product | EC_number | more information.
    + script: `tblToLst.awk` per genome.
    + output: `LSTINFO/<gembase_name>.lst`
- Copy `Database/<gembase_name>-gembase.fna` to `Replicons/<gembase_name.fna`
- Generate .prt (resp. .gen) fles, fromm .faa (resp .ffn) files generated by prokka in `<gembase_name>-prokka11Res`.
    + Script: `create_prt_gen.py` per genome
    + Output: `Genes/<gembase_name>.gen` and `Proteins/<gembase_name>.prt`.

Check that all went well (number of proteins/crispr/total genes corresponding between prokka results and generated files) with `check_gembase.sh`. If problems occur, they are added to the error output file, called `err-<LSTINFO_file>-<current_date>.err`.
